<a href="https://colab.research.google.com/github/jamesqiu2021/DSCI-401-QW/blob/main/HW3-Qiu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
airlines = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/airlines.csv")
flights = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/flights.csv")
airports = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/airports.csv")
batting = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/Batting.csv")
people = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/People.csv")
Pitching = pd.read_csv("https://raw.githubusercontent.com/jamesqiu2021/DSCI-401-QW/main/Pitching.csv")
planes = pd.read_csv("https://raw.githubusercontent.com/tidyverse/nycflights13/main/data-raw/planes.csv")


#1(a) (10 points) Name every player in baseball history who has accumulated at least 300 home runs (HR) and at least 300 stolen bases (SB).You can find the first and last name of the player in the People data frame. Join this to your result along with the total home runs and total bases stolen for each of these elite players.

In [14]:
from numpy.lib.shape_base import column_stack
column_stack(people)

array([['playerID', 'birthYear', 'birthMonth', 'birthDay',
        'birthCountry', 'birthState', 'birthCity', 'deathYear',
        'deathMonth', 'deathDay', 'deathCountry', 'deathState',
        'deathCity', 'nameFirst', 'nameLast', 'nameGiven', 'weight',
        'height', 'bats', 'throws', 'debut', 'finalGame', 'retroID',
        'bbrefID', 'deathDate', 'birthDate']], dtype='<U12')

In [15]:
from pandas.core.resample import PeriodIndexResampler
from pandas.core.reshape.merge import merge
q = batting.loc[:,["playerID","HR","SB"]].groupby("playerID").sum()
q1=q[q.HR>300][q.SB>300]
people["name"]=people["nameFirst"]+people["nameLast"]
people_1=people[["playerID","name"]]
merge_q = pd.merge(people_1,q1,on="playerID",how="inner")
merge_q




/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


playerID           name   HR     SB
0  beltrca01  CarlosBeltran  435  312.0
1  bondsba01     BarryBonds  762  514.0
2  bondsbo01     BobbyBonds  332  461.0
3  dawsoan01    AndreDawson  438  314.0
4  finlest01    SteveFinley  304  320.0
5   mayswi01     WillieMays  660  338.0
6  rodrial01  AlexRodriguez  696  329.0
7  sandere02  ReggieSanders  305  304.0

#1(b) (10 points) Similarly, name every pitcher in baseball history who has accumulated at least 300 wins (W) and at least 3,000 strikeouts (SO).

In [17]:
q2 = Pitching.loc[:,["playerID","W","SO"]].groupby("playerID").sum()
q3=q2[q2.W>300][q2.SO>3000]
people["name"]=people["nameFirst"]+people["nameLast"]
people_1=people[["playerID","name"]]
merge_q2 = pd.merge(people_1,q3,on="playerID",how="inner")
merge_q2

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


playerID           name    W    SO
0  carltst01   SteveCarlton  329  4136
1  clemero02   RogerClemens  354  4672
2  johnsra05   RandyJohnson  303  4875
3  johnswa01  WalterJohnson  417  3509
4  maddugr01     GregMaddux  355  3371
5  niekrph01     PhilNiekro  318  3342
6  perryga01   GaylordPerry  314  3534
7   ryanno01      NolanRyan  324  5714
8  seaveto01      TomSeaver  311  3640
9  suttodo01      DonSutton  324  3574

#1(c) (10 points) Identify the name and year of every player who has hit at least 50 home runs in a single season. Which player had the lowest batting average in that season?

In [ ]:
q1c1=batting[["yearID","playerID","HR"]][batting.HR>=50]
merge_q3 = pd.merge(q1c1,people_1,on="playerID",how="inner")
merge_q3


In [20]:
merge_q4 = pd.merge(merge_q3,batting,on=("playerID","yearID"),how="inner")
merge_q4["batting_average"]=merge_q4["H"]/merge_q4["AB"]
merge_q4[["yearID","playerID","name","batting_average"]].sort_values(by="batting_average",ascending=True).head(5)

yearID   playerID          name  batting_average
44    2019  alonspe01    PeteAlonso         0.259631
40    2010  bautijo02  JoseBautista         0.260105
36    2005  jonesan01   AndruwJones         0.262799
15    1961  marisro01    RogerMaris         0.269492
29    1998  vaughgr01    GregVaughn         0.272251

#2. Use the nycfights13 package and the fights and planes tables to answer the following questions:

#2(a) (10 points) What is the oldest plane (specified by the tailnum variable) that flew from New York City airports in 2013?

In [7]:
q2a_1 = flights[["tailnum","dep_time"]].dropna(subset=["dep_time"])
q2a_2 = pd.merge(q2a_1,planes,on="tailnum",how="inner")
q2a_2.sort_values(by="year",ascending=True).head(5)

tailnum  dep_time    year                     type manufacturer  \
264146  N381AA    1505.0  1956.0  Fixed wing multi engine      DOUGLAS   
264129  N381AA    1525.0  1956.0  Fixed wing multi engine      DOUGLAS   
264130  N381AA    1737.0  1956.0  Fixed wing multi engine      DOUGLAS   
264131  N381AA     817.0  1956.0  Fixed wing multi engine      DOUGLAS   
264132  N381AA    1528.0  1956.0  Fixed wing multi engine      DOUGLAS   

         model  engines  seats  speed         engine  
264146  DC-7BF        4    102  232.0  Reciprocating  
264129  DC-7BF        4    102  232.0  Reciprocating  
264130  DC-7BF        4    102  232.0  Reciprocating  
264131  DC-7BF        4    102  232.0  Reciprocating  
264132  DC-7BF        4    102  232.0  Reciprocating

#2(b) (10 points) How many airplanes that flew from New York City are included in the planes table?

In [41]:
q2b = set(q2a_2.tailnum)
len(q2b)

3316

In [42]:
planes["tailnum"].count()

3322

**# 3316 planes that flew from New York City are included in the planes table. **